README
	nokryptia v1.3

This program is made for the Nokia 5510 to put music on your
mobile.  To use the phone, 2 steps are necessary to do :


1. using the usb-storage :

include the following line into the kernel source in the file
/usr/src/linux/drivers/usb/storage/unusual_devs.h - assuming
that you installed the kernel source in /usr/src/linux

UNUSUAL_DEV(  0x0421, 0x0404, 0x0000, 0xffff,
		"Nokia",
		"5510",
		US_SC_SCSI, US_PR_BULK, NULL, US_FL_START_STOP),

--------------------------------------------------------------------
For people who want to understand this: the first two parameters
are the vendor and the product code.  This is easy to find, just
plug in your mobile and the kernel will report it.  The next two
arguments were just a guess for me, I don't know what they mean.
It must be a minimum and a maximum of something, other examples
in the same file also used 0x0000 and 0xffff.  The next two
parameters are vendor and product string.  Feel free to change
this in wathever you like (I never tested this, but I think it
works with any string).  The next parameter tells the driver to
use SCSI commands for the device.  The engineers who invented
usb-storage took definitely scsi as an example.  The next parameter
is the protocol, I can't tell very much about this.  The next one
is NULL, this should be a pointer to an initialization function.
Luckily, there is no need for this on the Nokia, but this was
easy to guess because even windows didn't need drivers to use the
storage function.  The last one is probably the most important
parameter, these are the flags.  In this case, we need only one
for telling the driver that we need to ignore the start and stop
commands.  For this reason, the patch is not necessary anymore in
kernels >= 2.5.48, because these kernels will never do start and
stop commands, except if the device is asking for it.
--------------------------------------------------------------------

After recompiling the kernel you should be able to mount the
device as /dev/sda1.  I tested this with kernel version 2.4.18,
which seems to work; it is also reported to work with kernel
version 2.4.19  Probably it works also on earlier kernel
versions, as long as they support usb-storage.  I only tested
my personal 5510, so it may not work with your phone.
WARNING : before unmounting your mobile, do a sync.  I noticed
that you will get some ugly errors from the kernel if you forget
this.  The sync can take up some minutes, depending on how much
data will be stored.  On some distro's the sync seems not to be
necessary.
Some research learned me that this patch is not necessary anymore
for kernel versions > 2.5.48.  Because this is almost the latest
development version, I suppose that you will be interested in
the patch anyway.
I got a lot of response from people who were not familiar with
kernel recompiling.  Because kernel-recompiling is certainly not
easy, I will give some useful hints to avoid answering always
the same questions via e-mail.  If you still have problems, feel
free to contact me, it could be that I forgot to add a very
important hint in the list.

a) Be sure to compile at least the usb-storage, the scsi_mod,
   the sd_mod, the vfat and the fat modules.  Of course, you will
   need more modules, depending of the rest of your hardware.
   Keep in mind that these should be modules, I tried to compile
   it built-in into the kernel and it didn't work.
b) On some distributions, it seems to be necessary to do
   "modprobe usb-storage" after plugging in your mobile.  This
   was the most common solution to the problems sent to me in the
   past, so be sure to check especially this.
c) The device that needs to be mounted is /dev/sda1.  If you
   have scsi-harddisk it will probably be something else like
   /dev/sdb1 or /dev/sdc1.  If it happens to be that you have a
   scsi-cdrom, the cdrom will be /dev/scd0 and the mobile is still
   /dev/sda1.  If this really doesn't work, the kernel messages
   produced after plugging in your phone can give you some hints.
d) If it happens that /dev/sda1 does not exist on your system, you
   will need to do a mknod to make the node.  As I don't know how
   to do this, consult the man-page of mknod.  The output you
   should get when doing "ls -l /dev/sda1" is:
   brw-rw----    1 root     disk       8,   1 Mar 24  2001 /dev/sda1


2. putting music on your mobile :

Normally, the music on the mobile must be encrypted using the
Nokia Audio Manager.  Because I don't know the encryption 
algorithm, I made a workaround by setting the key to all zeroes
and using the unencrypted music, which seems to work on my phone.
The program makes use of id3lib.  Install this before you install
nokryptia.  You can find it at http://id3lib.sourceforge.net.
To compile, enter the following commands :

% tar xvzf nokryptia-1.3.tgz
% cd nokryptia-1.3
% ./configure
% make
(finally, as root)
% make install

To put the music on your device, change the working directory to
the place where the device is mounted (eg /mnt/nokia).  Now run
the utility with an mp3 file as argument.

Enjoy your life without window$!

Roel Derickx
E-mail : roel.derickx@campux.be



ChangeLog
22 december 2003
- include option to specify output directory
- public release of version 1.3

05 december 2003
- updated version of webpage is available

03 december 2003
- fixed lse to mp3 conversion again to make it really work

02 december 2003
- code clean-up: make use of getopt function calls (thanks to Marian Hromiak)

19 july     2003
- fixed bug that prevented conversion from lse back to mp3 (how can 
  such a bug live that long without being noticed?)
- verbose option is giving more info
- some simplifications of the code

11 may      2003
- fixed bug that prevented -d and -u from being the second parameter
- fixed string overflow at ID3tags with artist or songtitle > 30 chars
		   
24 february 2003
- mobilix is changed into tuxmobil.org

18 december 2003
- first version of webpage is ready

15 december 2002
- namechange to avoid copyright troubles
- add manual-page, add extra information to README based on problems
  that were send to me.
      
11 december 2002
- added ID3 tag support (thanks to Krzysztof Gorgolewski)

09 december 2002
- first public release
  address is http://www.mobilix.org/

08 december 2002
- fixed some memory allocation bugs
- outputfile is now written to the working directory
- add verbose option

07 december 2002
- initial version
- include parseOptions() to get options
- switched artistname and songtitle


src/config.h
/* src/config.h.  Generated by configure.  */
/* src/config.h.in.  Generated from configure.in by autoheader.  */

/* Define to 1 if you have the `bcopy' function. */
#define HAVE_BCOPY 1

/* Define to 1 if you have the <inttypes.h> header file. */
#define HAVE_INTTYPES_H 1

/* Define to 1 if you have the <iostream.h> header file. */
/* #undef HAVE_IOSTREAM_H */

/* Define to 1 if you have the `id3' library (-lid3). */
#define HAVE_LIBID3 1

/* Define to 1 if you have the `z' library (-lz). */
#define HAVE_LIBZ 1

/* Define to 1 if you have the <memory.h> header file. */
#define HAVE_MEMORY_H 1

/* Define to 1 if you have the <stdint.h> header file. */
#define HAVE_STDINT_H 1

/* Define to 1 if you have the <stdio.h> header file. */
#define HAVE_STDIO_H 1

/* Define to 1 if you have the <stdlib.h> header file. */
#define HAVE_STDLIB_H 1

/* Define to 1 if you have the `strcpy' function. */
#define HAVE_STRCPY 1

/* Define to 1 if you have the <strings.h> header file. */
#define HAVE_STRINGS_H 1

/* Define to 1 if you have the <string.h> header file. */
#define HAVE_STRING_H 1

/* Define to 1 if you have the `strncpy' function. */
#define HAVE_STRNCPY 1

/* Define to 1 if you have the `strstr' function. */
#define HAVE_STRSTR 1

/* Define to 1 if you have the <sys/stat.h> header file. */
#define HAVE_SYS_STAT_H 1

/* Define to 1 if you have the <sys/types.h> header file. */
#define HAVE_SYS_TYPES_H 1

/* Define to 1 if you have the <unistd.h> header file. */
#define HAVE_UNISTD_H 1

/* Define to 1 if your C compiler doesn't accept -c and -o together. */
/* #undef NO_MINUS_C_MINUS_O */

/* Name of package */
#define PACKAGE "nokryptia"

/* Define to the address where bug reports for this package should be sent. */
#define PACKAGE_BUGREPORT ""

/* Define to the full name of this package. */
#define PACKAGE_NAME ""

/* Define to the full name and version of this package. */
#define PACKAGE_STRING ""

/* Define to the one symbol short name of this package. */
#define PACKAGE_TARNAME ""

/* Define to the version of this package. */
#define PACKAGE_VERSION ""

/* Define to 1 if you have the ANSI C header files. */
#define STDC_HEADERS 1

/* Version number of package */
#define VERSION "1.3"


nokryptia.cpp
/*
 * nokryptia : this program is a workaround for the encryption used
 *             in the mp3 player included with nokia5510 mobiles.
 *
 * Copyright (C) 2002 Roel Derickx <roel.derickx@campux.be>
 * Copyright (C) 2020 Charlene Wendling <julianaito@posteo.jp>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */
 
#include <stdio.h>
#include <iostream>
#include <string.h>
#include <getopt.h>
#include <id3/id3lib_streams.h>
#include <id3/tag.h>

#define COPYBUFSIZE 10240

FILE * infile;
FILE * outfile;

char * infilename;
char * outfilename;
char * artistname;
char * songtitle;
int    verbose;
int    action;     /* 1 for upload, 2 for download */

void writeHelpMessage(const char * progname) {
	std::cout << PACKAGE << " version " << VERSION <<
	    " copyright by Roel Derickx and others\n"
	    "THIS SOFTWARE COMES WITH ABSOLUTELY NO WARRANTY! "
	    "USE AT YOUR OWN RISK!\n"
	    "\nUsage:\n"
	    "   " << progname << " [options] file\n"
	    "\nOptions:\n"
	    "   -u               convert mp3 into lse for upload\n"
	    "   -d               convert lse into mp3 for download\n"
	    "   -a <artistname>  set artist to artistname\n"
	    "   -t <songtitle>   set title to songtitle\n"
	    "   -o <directory>   directory to where output files should be written\n"
	    "   -v               be verbose\n"
	    "   -h               print this help screen\n"
	    "\nSee the man page for more information\n";
}

/*
 * looknchange
 * This function looks for a pattern string2 in string1. If found,
 * string3 is placed in string1 on the place where string2 was found,
 * and the return value will be 0. If string2 was not found, return
 * value is non-zero.
 */
int looknchange(char * string1, char * string2, char * string3) {
	char *tmp;
	if (strlen(string1) < strlen(string2)) 
		return 1;
	tmp = string1+strlen(string1)-strlen(string2);
	if (strcasecmp(tmp, string2) != 0)
		return 1;
	strcpy(tmp, string3);
	return 0;
}

int parseInputFiles(int argc, char *argv[]){
	char * temp;

	/* look for .mp3 in infilename and change into .lse */
	/* or vice versa, depending on which action is chosen */
	/* also remove the path from infilename */

	if (optind >= argc) {
		std::cout << "No input file given!" << std::endl;
		return 1;
	}
	strcpy(infilename, argv[optind++]);
	if ((outfilename[strlen(outfilename)] == '/') ||
	    (strlen(outfilename) == 0)) {
		strcat(outfilename,"\0");
	}
	else {
		strcat(outfilename,"/\0");
	}
	temp = rindex(infilename,'/');
	if (temp != NULL)
		strcat(outfilename, temp+1);
	else
		strcat(outfilename, infilename);
  
  
	switch (action) {
	case 1:
		if (looknchange(outfilename, (char *) ".mp3", (char *) ".lse") != 0) {
			std::cout << "Input file must have mp3 "
			     << "extension!" << std::endl;
			return 1;
		}
		break;
	case 2:
		if (looknchange(outfilename, (char *) ".lse", (char *) ".mp3") != 0) {
			std::cout << "Input file must have lse "
			     << "extension!" << std::endl;
			return 1;
		}
		break;
	} /* end of switch */

	return 0;
}

int parseOptions(int argc, char * argv[]) {
	int c;
	
	/* first of all, check number of parameters */
	if (argc < 2) {
		return 1;
	}

	/* set default settings */
	strcpy(artistname,"unknown");
	strcpy(songtitle,"unknown");
	strcpy(infilename,"unknown");
	strcpy(outfilename,"\0");
	verbose = 0;
	action = 1; /* default action = upload */

	/* loop through all parameters */
	while ((c=getopt(argc,argv,"uda:t:hvo:"))!= -1 ){
		switch (c) {
		case 'h':
			writeHelpMessage(argv[0]);
			exit(0);
			break;
		case 'a':
			strncpy(artistname,optarg,30);
			strcat(artistname,"\0");
			break;
		case 't':
			strncpy(songtitle,optarg,30);
			strcat(songtitle,"\0");
			break;
		case 'v':
			verbose=1;
			break;
		case 'u':
			action=1;
			break;
		case 'd':
			action=2;
			break;
		case 'o':
			strcpy(outfilename,optarg);
			strcat(outfilename,"\0");
			break;
		default:
			std::cout << "Unknown option " << optopt << std::endl;
		} /* end of switch */
	} /* end of while */

	/* id3 tag support */
	ID3_Tag myTag;
	myTag.Link(infilename);
	ID3_Frame * myFrame;
	
	if ((strcmp(songtitle,"unknown") == 0) &&
	    (myFrame = myTag.Find(ID3FID_TITLE))) {
		myFrame->Field(ID3FN_TEXT).Get(songtitle,30);
		strcat(songtitle,"\0");
	}
	
	if ((strcmp(artistname,"unknown") == 0) &&
	    (myFrame = myTag.Find(ID3FID_LEADARTIST))) {
		myFrame->Field(ID3FN_TEXT).Get(artistname,30);
		strcat(artistname,"\0");
	}

	return parseInputFiles(argc,argv);
}

void makeTag(char * artist, char * title) {
	char * buf;
	int lengthArtist = strlen(artist);
	int lengthTitle  = strlen(title);

	if (verbose) {
		std::cout << "Writing tag to output :" << std::endl;
		std::cout << "artistname = " << artist << std::endl;
		std::cout << "songtitle  = " << title << std::endl;
	}

	/* allocate memory */
	buf = new char[384]();

	/* generate the data */
	memcpy(buf,"ID3",3);
	*(buf+3) = 0x03;
	*(buf+8) = 0x02;
	memcpy(buf+9,"vTIT2",5);
	*(buf+17) = lengthTitle+1;
	memcpy(buf+21,title,lengthTitle);
	memcpy(buf+21+lengthTitle,"TPE1",4);
	*(buf+28+lengthTitle) = lengthArtist+1;
	memcpy(buf+32+lengthTitle,artist, lengthArtist);
	memcpy(buf+32+lengthTitle+lengthArtist,"TLEN",4);
	*(buf+39+lengthTitle+lengthArtist) = 0x06;

	/* write the header to the file */
	fwrite(buf,384,1,outfile);

	/* clean up the memory */
	delete[] buf;
}

void makeHeader() {
	char * buf;

	if (verbose) std::cout << "Writing LockStream key to output" << std::endl;

	/* allocate memory */
	buf = new char[512]();

	/* generate the data */
	memcpy(buf,"LockStream Embedded\0d",21);
	*(buf+25) = 0x02;
	*(buf+32) = 0x02;
	*(buf+37) = 0x02;
	*(buf+104) = 0x02;
	*(buf+108) = 0x13;
	*(buf+112) = 0x80;

	/* write the header to the file */
	fwrite(buf,512,1,outfile);

	/* clean up the memory */
	delete[] buf;
}

int readAndTestHeader() {
	int bread, flag;
	char * buf;

	if (verbose) {
		std::cout << "Testing if the lse file is not written by "
		     << "the Nokia Audio Manager" << std::endl;
	}
	
	/* allocate memory */
	/* 384 bytes for the tag, 512 for the lockstream header */
	buf = new char[896];

	bread = fread(buf,1,896,infile);
	if (bread != 896) {
		std::cout << "Input file is too small!" << std::endl;
		return 0;
	}
	
	/* if this file is made by the official nokia program, */
	/* the key is not all zeroes and the file is thus encrypted */
	flag = 1;
	for (int i = 422; i < 526; i++) {
		if ((*(buf+i) != 0) &&
		    (i != 488) &&
		    (i != 492) &&
		    (i != 496)) flag = 0;
	}

	/* clean up the memory */
	delete[] buf;

	/* test is done, return value of flag */
	return flag;
}

void copyData() {
	int bread, bwritten;
	char * buf;

	if (verbose) std::cout << "Writing data to output" << std::endl;

	/* allocate memory */
	buf = new char[COPYBUFSIZE];
	
	/* copy the original mp3 into the lse */
	do {
		bread = fread(buf,1,COPYBUFSIZE,infile);
		bwritten = fwrite(buf,1,bread,outfile);
	} while (bread == bwritten && feof(infile) == 0);

	/* clean up the memory */
	delete[] buf;
}

void allocateMemory() {
	/* allocate memory for global variables */
	artistname = new char[31];
	songtitle = new char[31];
	infilename = new char[1000];
	outfilename = new char[1000];
}

void cleanUpMemory() {
	/* clean up memory */
	delete [] outfilename;
	delete [] infilename;
	delete [] songtitle;
	delete [] artistname;
}

int main(int argc, char * argv[]) {
	allocateMemory();

	/* parse the options */
	if (parseOptions(argc, argv) != 0) {
		writeHelpMessage(argv[0]);
		
		cleanUpMemory();
		
		/* exit with errorcode 1 */
		return 1;
	}
	
	/* open the infile */
	if ((infile = fopen(infilename,"rb")) == NULL) {
		std::cout << infilename
		     << ": No such file or directory" << std::endl;
		
		cleanUpMemory();
		
		/* exit with errorcode 1 */
		return 1;
	}
	
	/* open the outfile */
	if ((outfile = fopen(outfilename,"wb")) == NULL) {
		std::cout << "An error occured while opening "
		     << outfilename << " for writing." << std::endl
		     << "Do you have write permission "
		     << "in this directory?" << std::endl;
		
		cleanUpMemory();
		
		/* close the infile */
		fclose(infile);
		
		/* exit with errorcode 1 */
		return 1;
	}
	
	/* do the actual job */
	if (action == 1) {
		makeTag(artistname, songtitle);
		makeHeader();
		copyData();
	}
	else if (action == 2) {
		if (readAndTestHeader()) {
			copyData();
		}
		else {
			std::cout << "This file was encrypted by the "
			     << "official nokia program." << std::endl
			     << "Unable to decrypt!" << std::endl;
		}
	}
	
	/* close files */
	fclose(outfile);
	fclose(infile);

	cleanUpMemory();
	
	/* exit the program cleanly */
	return 0;
}



src/nokryptia.cc
/*
 * nokryptia : this program is a workaround for the encryption used
 *             in the mp3 player included with nokia5510 mobiles.
 *
 * Copyright (C) 2002 Roel Derickx <roel.derickx@campux.be>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */
 
#include <stdio.h>
#include <iostream>
#include <string.h>
#include <getopt.h>
#include "config.h"
#include "id3/id3lib_streams.h"
#include "id3/tag.h"

#define COPYBUFSIZE 10240

FILE * infile;
FILE * outfile;

char * infilename;
char * outfilename;
char * artistname;
char * songtitle;
int    verbose;
int    action;     /* 1 for upload, 2 for download */

void writeHelpMessage(const char * progname) {
	std::cout << PACKAGE << " version " << VERSION
	     << " copyright by Roel Derickx" << std::endl
	     << "THIS SOFTWARE COMES WITH ABSOLUTELY NO WARRANTY! "
	     << "USE AT YOUR OWN RISK!" << std::endl << std::endl
	     << "Usage:" << std::endl
	     << "   " << progname << " [options] file" << std::endl
	     << "Options:" << std::endl
	     << "   -u               convert mp3 into lse for upload" << std::endl
	     << "   -d               convert lse into mp3 for download" << std::endl
	     << "   -a <artistname>  set artist to artistname" << std::endl
	     << "   -t <songtitle>   set title to songtitle" << std::endl
	     << "   -o <directory>   directory to where output files should be written" << std::endl
	     << "   -v               be verbose" << std::endl
	     << "   -h               print this help screen" << std::endl
	     << "See the man page for more information" << std::endl;
}

/*
 * looknchange
 * This function looks for a pattern string2 in string1. If found,
 * string3 is placed in string1 on the place where string2 was found,
 * and the return value will be 0. If string2 was not found, return
 * value is non-zero.
 */
int looknchange(char * string1, char * string2, char * string3) {
	char *tmp;
	if (strlen(string1) < strlen(string2)) 
		return 1;
	tmp = string1+strlen(string1)-strlen(string2);
	if (strcasecmp(tmp, string2) != 0)
		return 1;
	strcpy(tmp, string3);
	return 0;
}

int parseInputFiles(int argc, char *argv[]){
	char * temp;
  
	/* look for .mp3 in infilename and change into .lse */
	/* or vice versa, depending on which action is chosen */
	/* also remove the path from infilename */

	if (optind >= argc) {
        std::cout << "No input file given!" << std::endl;
		return 1;
	}
	strcpy(infilename, argv[optind++]);
	if ((outfilename[strlen(outfilename)] == '/') ||
	    (strlen(outfilename) == 0)) {
		strcat(outfilename,"\0");
	}
	else {
		strcat(outfilename,"/\0");
	}
	temp = rindex(infilename,'/');
	if (temp != NULL)
		strcat(outfilename, temp+1);
	else
		strcat(outfilename, infilename);
  
  
	switch (action) {
	case 1:
		if (looknchange(outfilename,".mp3",".lse") != 0) {
            std::cout << "Input file must have mp3 "
			     << "extension!" << std::endl;
			return 1;
		}
		break;
	case 2:
		if (looknchange(outfilename,".lse",".mp3") != 0) {
            std::cout << "Input file must have lse "
			     << "extension!" << std::endl;
			return 1;
		}
		break;
	} /* end of switch */

	return 0;
}

int parseOptions(int argc, char * argv[]) {
	int i,c;
	char * temp;
	
	/* first of all, check number of parameters */
	if (argc < 2) {
		return 1;
	}

	/* set default settings */
	strcpy(artistname,"unknown");
	strcpy(songtitle,"unknown");
	strcpy(infilename,"unknown");
	strcpy(outfilename,"\0");
	verbose = 0;
	action = 1; /* default action = upload */

	/* loop through all parameters */
	while ((c=getopt(argc,argv,"uda:t:hvo:"))!= -1 ){
		switch (c) {
		case 'h':
			writeHelpMessage(argv[0]);
			exit(0);          
			break;
		case 'a':
			strncpy(artistname,optarg,30);
			strcat(artistname,"\0");
			break;
		case 't':
			strncpy(songtitle,optarg,30);
			strcat(songtitle,"\0");
			break;
		case 'v':
			verbose=1;
			break;
		case 'u':
			action=1;
			break;
		case 'd':
			action=2;
			break;
		case 'o':
			strcpy(outfilename,optarg);
			strcat(outfilename,"\0");
			break;
		default:
			std::cout << "Unknown option " << optopt << std::endl;
		} /* end of switch */
	} /* end of while */

	/* id3 tag support */
	ID3_Tag myTag;
	myTag.Link(infilename);
	ID3_Frame * myFrame;
	
	if ((strcmp(songtitle,"unknown") == 0) &&
	    (myFrame = myTag.Find(ID3FID_TITLE))) {
		myFrame->Field(ID3FN_TEXT).Get(songtitle,30);
		strcat(songtitle,"\0");
	}
	
	if ((strcmp(artistname,"unknown") == 0) &&
	    (myFrame = myTag.Find(ID3FID_LEADARTIST))) {
		myFrame->Field(ID3FN_TEXT).Get(artistname,30);
		strcat(artistname,"\0");
	}
		
	return parseInputFiles(argc,argv);
}

void makeTag(char * artist, char * title) {
	char * buf;
	int i;
	int lengthArtist = strlen(artist);
	int lengthTitle  = strlen(title);

	if (verbose) {
		std::cout << "Writing tag to output :" << std::endl;
		std::cout << "artistname = " << artist << std::endl;
		std::cout << "songtitle  = " << title << std::endl;
	}

	/* allocate memory */
	buf = new char[384];

	/* clear the buffer */
	for (i = 0; i < 384; i++) {
		*(buf+i) = 0;
	}

	/* generate the data */
	bcopy("ID3",buf,3);
	*(buf+3) = 0x03;
	*(buf+8) = 0x02;
	bcopy("vTIT2",buf+9,5);
	*(buf+17) = lengthTitle+1;
	bcopy(title, buf+21, lengthTitle);
	bcopy("TPE1", buf+21+lengthTitle, 4);
	*(buf+28+lengthTitle) = lengthArtist+1;
	bcopy(artist, buf+32+lengthTitle, lengthArtist);
	bcopy("TLEN", buf+32+lengthTitle+lengthArtist, 4);
	*(buf+39+lengthTitle+lengthArtist) = 0x06;

	/* write the header to the file */
	fwrite(buf,384,1,outfile);

	/* clean up the memory */
	delete[] buf;
}

void makeHeader() {
	char * buf;
	int i;

	if (verbose) std::cout << "Writing LockStream key to output" << std::endl;

	/* allocate memory */
	buf = new char[512];

	/* clear the buffer */
	for (i = 0; i < 512; i++) {
		*(buf+i) = 0;
	}

	/* generate the data */
	bcopy("LockStream Embedded\0d",buf,21);
	*(buf+25) = 0x02;
	*(buf+32) = 0x02;
	*(buf+37) = 0x02;
	*(buf+104) = 0x02;
	*(buf+108) = 0x13;
	*(buf+112) = 0x80;

	/* write the header to the file */
	fwrite(buf,512,1,outfile);

	/* clean up the memory */
	delete[] buf;
}

int readAndTestHeader() {
	int bread, flag;
	char * buf;

	if (verbose) {
		std::cout << "Testing if the lse file is not written by "
		     << "the Nokia Audio Manager" << std::endl;
	}
	
	/* allocate memory */
	/* 384 bytes for the tag, 512 for the lockstream header */
	buf = new char[896];

	bread = fread(buf,1,896,infile);
	if (bread != 896) {
		std::cout << "Input file is too small!" << std::endl;
		return 0;
	}
	
	/* if this file is made by the official nokia program, */
	/* the key is not all zeroes and the file is thus encrypted */
	flag = 1;
	for (int i = 422; i < 526; i++) {
		if ((*(buf+i) != 0) &&
		    (i != 488) &&
		    (i != 492) &&
		    (i != 496)) flag = 0;
	}

	/* clean up the memory */
	delete[] buf;

	/* test is done, return value of flag */
	return flag;
}

void copyData() {
	int bread, bwritten;
	char * buf;

	if (verbose) std::cout << "Writing data to output" << std::endl;

	/* allocate memory */
	buf = new char[COPYBUFSIZE];
	
	/* copy the original mp3 into the lse */
	do {
		bread = fread(buf,1,COPYBUFSIZE,infile);
		bwritten = fwrite(buf,1,bread,outfile);
	} while (bread == bwritten && feof(infile) == 0);

	/* clean up the memory */
	delete[] buf;
}

void allocateMemory() {
	/* allocate memory for global variables */
	artistname = new char[31];
	songtitle = new char[31];
	infilename = new char[1000];
	outfilename = new char[1000];
}

void cleanUpMemory() {
	/* clean up memory */
	delete [] outfilename;
	delete [] infilename;
	delete [] songtitle;
	delete [] artistname;
}

int main(int argc, char * argv[]) {
	allocateMemory();

	/* parse the options */
	if (parseOptions(argc, argv) != 0) {
		writeHelpMessage(argv[0]);
		
		cleanUpMemory();
		
		/* exit with errorcode 1 */
		return 1;
	}
	
	/* open the infile */
	if ((infile = fopen(infilename,"rb")) == NULL) {
		std::cout << infilename 
		     << ": No such file or directory" << std::endl;
		
		cleanUpMemory();
		
		/* exit with errorcode 1 */
		return 1;
	}
	
	/* open the outfile */
	if ((outfile = fopen(outfilename,"wb")) == NULL) {
		std::cout << "An error occured while opening " 
		     << outfilename << " for writing." << std::endl
		     << "Do you have write permission "
		     << "in this directory?" << std::endl;
		
		cleanUpMemory();
		
		/* close the infile */
		fclose(infile);
		
		/* exit with errorcode 1 */
		return 1;
	}
	
	/* do the actual job */
	if (action == 1) {
		makeTag(artistname, songtitle);
		makeHeader();
		copyData();
	}
	else if (action == 2) {
		if (readAndTestHeader()) {
			copyData();
		}
		else {
			std::cout << "This file was encrypted by the "
			     << "official nokia program." << std::endl
			     << "Unable to decrypt!" << std::endl;
		}
	}
	
	/* close files */
	fclose(outfile);
	fclose(infile);

	cleanUpMemory();
	
	/* exit the program cleanly */
	return 0;
}

